### PR TITLE
Gutenlypso: update calypso classic block to update block content even if the editor instance is marked for removal

### LINKED ISSUE #29643: More Tag option not working on Classic block
#### Steps to reproduce
1. Add a Classic block 
2. Add the "more tag" between 2 paragraphs
3. Save and Update
4. Check post from Blog page, the tag is not applied
5. Check post individually, there is a broken "read more" reference (screenshot below)

#### What I expected
The more tag should be added and should be working (testing from an archive page or blog page)

#### What happened instead

The more tag is not added, not working, and there is a broken "read more" reference  as seen in the following  screenshot:
screen shot 2018-12-19 at 18 45 16

### PR DESCRIPTION
**Pull request by gwwar**
#### Changes proposed in this Pull Request

I don't really understand why the timing of the tinymce lifecycle events are lining up this way, but once we reach handlers like the ones that manipulate the more tag, the editor is already marked for removal, and so any content manipulation is not reflected. 

In this PR I explicitly detect for this case, then update block content. I've also modified some behavior to more closely match the core Classic block. Basically we don't bother to update block content until on block blur rather than a debounced onchange event.

In the case of the more tag, we should see a nice visual placeholder in the visual mode, and an html more comment in the html view.

**Before**:
broken

**After**:
works

#### Testing instructions

**Publishing with a more block renders correctly**
* Add a Classic block
* Add the "more tag" between 2 paragraphs
* Save and Update
* Check post from Blog page, the tag is not applied
* Check post individually, more tag works as expected

**Visually toggling between visual/html shows the correct view**
* Add a Classic block
* Add the "more tag" between 2 paragraphs
* Select block options, and pick "Edit As HTML"
works

**Converting to Blocks works**
* Add a Classic block
* Add the "more tag" between 2 paragraphs
* Select block options, and pick "Convert to Blocks"

convert to blocks

**Content is saved when we focus out of the Classic block**
This matches behavior in the core classic block.
* Publish a post with a classic block
* Open the post and make sure there are no unsaved changes on the draft
* Start typing in the Classic block, wait.
* Notice that the update option doesn't appear
* Blur the classic block by clicking outside of the area
* We should see an update option

Fixes #29643

### CODE REVIEW

**Comment by gwwar** [comment]
I need to update block content on blur to match core Classic block behavior, will circle back to this

```
		editor.on( 'blur', () => {
			bookmark = editor.selection.getBookmark( 2, true );

			setAttributes( {
				content: editor.getContent(),
			} );

			editor.once( 'focus', () => {
				if ( bookmark ) {
					editor.selection.moveToBookmark( bookmark );
				}
			} );

			return false;
		} );
```

**Comment by gwwar** [comment]
Okay this is all set for review. @jsnajdr it looks like icfy might need some 👀 seeing a `curl: (7) Failed to connect to api.iscalypsofastyet.com port 5000: Connection refused` in the failed run.

**Comment by gwwar** [comment]
Thanks @jsnajdr !

**Comment by gwwar** [comment]
This PR should also fix https://github.com/Automattic/wp-calypso/issues/29833

emoji fix

File: client/gutenberg/extensions/classic-block/edit.jsx [hunk 1]
```diff
@@ -23,22 +23,56 @@ export class ClassicEdit extends Component {
 	}
 
 	storeEditor = ref => {
+		//By the time editor lifecycle events like PostProcess fire, this editor instance is already marked for removal
+		//so any content manipulation doesn't get reflected in this block.
+		//
+		//For now, when we detect an unmount, get content from the editor marked to be destroyed and update block attributes.
+		//https://github.com/Automattic/wp-calypso/issues/29643
```
Comment (dmsnell) [inline client/gutenberg/extensions/classic-block/edit.jsx]: small nitpick, but I'm used to seeing a space after `//` and before the text. are you comfortable adding one in on these lines?

other than that I like to use these comments as examples to others of how we can help other developers or our future selves by explaining what isn't obvious from the code itself - bravo

Comment (gwwar) [reply client/gutenberg/extensions/classic-block/edit.jsx]: Sure, I can update! I'm hoping this patch is a temporary fix until we get a better handle on the intended lifecycle loops.

File: client/gutenberg/extensions/classic-block/edit.jsx [hunk 2]
```diff
@@ -23,22 +23,56 @@ export class ClassicEdit extends Component {
...
+		if ( ! ref && this.editor ) {
```
Comment (dmsnell) [inline client/gutenberg/extensions/classic-block/edit.jsx]: to be more explicit it could help to name this value. I think it doesn't take much time to figure out but that's partially because I know what to look for from the PR description…

```js
const isUnmounting = ! ref && this.editor;

if ( isUnmounting ) {
	…
}
```

File: client/gutenberg/extensions/classic-block/edit.jsx [hunk 3]
```diff
@@ -23,22 +23,56 @@ export class ClassicEdit extends Component {
...
+		}
+	}
+
+	updateBlockContentAndBookmark = () => {
+		if ( ! this.editor ) {
+			return;
+		}
+		const editor = this.editor._editor;
+
+		this.bookmark = editor.selection.getBookmark( 2, true );
```
Comment (dmsnell) [inline client/gutenberg/extensions/classic-block/edit.jsx]: can we explain the parameters here since it'd be hard to follow `getBookmark` to see its source?

Comment (gwwar) [reply client/gutenberg/extensions/classic-block/edit.jsx]: Sure, it's a copy pasta from the core classic block, of which I had to dig into source to see the actual value. It returns an offsetBookmark. I'll add a comment.
 
https://www.tiny.cloud/docs/api/tinymce.dom/tinymce.dom.bookmarkmanager/#getbookmarkhttps://github.com/tinymce/tinymce/blob/146c3c7c98206f21fbe12f61537ff42dbe5c6c3a/src/core/main/ts/bookmark/GetBookmark.ts#L227
https://github.com/tinymce/tinymce/blob/146c3c7c98206f21fbe12f61537ff42dbe5c6c3a/src/core/main/ts/bookmark/GetBookmark.ts#L154

File: client/gutenberg/extensions/classic-block/edit.jsx (hunk 1 above)
Comment (gwwar) [reply client/gutenberg/extensions/classic-block/edit.jsx]: Updated in https://github.com/Automattic/wp-calypso/pull/29858/commits/108abb15de335f54ec4c9f45edfb6f52416e3329

File: client/gutenberg/extensions/classic-block/edit.jsx (hunk 3 above)
Comment (gwwar) [reply client/gutenberg/extensions/classic-block/edit.jsx]: Updated in 052ffe1

File: client/gutenberg/extensions/classic-block/edit.jsx (hunk 2 above)
Comment (gwwar) [reply client/gutenberg/extensions/classic-block/edit.jsx]: Good call. Added in 052ffe1

**Comment by mmtr** [comment]
I noticed a weird behavior after converting a classic block to blocks: if we insert new classic blocks, their content is not published.

These are the steps I follow for reproducing the issue:
- Add a classic block and enter some content on it
- Publish the post and verify that the content is published
- Convert the classic blocks
- Update the post and verify that the content is still on the published post
- Add a new classic block and enter some content on it
- Update the post (after blurring the classic block)

⚠️The published post doesn't contain the content of the new classic block

**Comment by gwwar** [comment]
Hmm @mmtr I can't reproduce this, but I can add back the onChange handling. I wouldn't be surprised if timing issues are causing this.

**Comment by kwight** [comment]
Works great for fixing #29643 and #29833 ✨ 

I was able to replicate @mmtr 's experience, but a further refresh and Update click in the editor corrected it. Looking at the API POST calls for the content updates, they both include the new content, so it does seem UI related.

**Comment by gwwar** [comment]
😭 alright I think I have most major cases. Note that when we go into "Edit as html" we're using the editor package's component `block-list/block-html.js` instead of this tinymce instance in the html tab.

**Comment by mmtr** [comment]
Thanks for fixing the issue with the blur events!

Now I have a different one when I switch from "Edit as HTML" to "Edit visually": changes done in the HTML editor are not reflected in the visual editor.

jan-04-2019 12-15-56

**Comment by gwwar** [comment]
Thanks @mmtr I'll maybe timebox debugging for that case, as adding content without blurring > convert to blocks works in visual mode. We can file a follow up issue if I'm unable to get it today.

fixing

**Comment by gwwar** [comment]
🤔 @mmtr are there more steps? I think I reproduced once, but now I'm not having luck. Is this pretty consistent for you?

**Comment by mmtr** [comment]
Uhm... strange. I cannot reproduce it now neither. But those were the only steps I followed. In fact, I recorded a couple more GIFs before that one, and I always did the same, so I don't know what can cause that state.

Will apply again the whole testing instructions and see if I detect how to reproduce it. Otherwise I think this is good to go.

**Comment by mmtr** [comment]
Same, I cannot reproduce it again, so I don't know what I did differently before. I guess we can ignore that since it should be a very corner case.

**Comment by gwwar** [comment]
Thanks for the detailed testing @mmtr and @kwight! I'll try and land this if I can get a clean full e2e run.

**Comment by gwwar** [comment]
I think I'm in the clear for the full e2e suite, I'll try landing and will revert if we note new breakages in master
